[FSDP2] support torch._foreach_copy_(float8) for fully_shard(Float8Linear)

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135955

this PR unblocks unit test with single Float8Linear module. It fixes following error
```
torch._foreach_copy_(foreach_copy_dsts, all_gather_inputs)
[rank0]:E0913 13:44:29.829000 2179476 torch/testing/_internal/common_distributed.py:671] RuntimeError: "foreach_tensor_copy" not implemented for 'Float8_e4m3fn'
```


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D63961071](https://our.internmc.facebook.com/intern/diff/D63961071)

weifengpy committed:
[FSDP2] all-gather in uint8 for float8-only modules

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Skylion007 commented on torch/distributed/_composable/fsdp/_fsdp_collectives.py:
```diff
@@ -442,6 +442,8 @@ def _get_all_gather_input_metadatas(
             input_numels.append(all_gather_input.numel())
         param_all_gather_input_dtypes.append(input_dtypes)
         param_all_gather_input_numels.append(input_numels)
+    if all_gather_dtype is torch.float8_e4m3fn:
```
Should this be for both float8 dtypes?

weifengpy replied:
good question! In total we have 4 float8 dtypes: https://github.com/pytorch/pytorch/blob/21ffa18ad1ad0bf22df8d7f42309190b4f3cc547/torch/storage.py#L524-L531

@vkuzo @drisspg could I assume parameters are always `float8_e4m3fn`? or should I list 4 float8 dtypes here?

@awgu this PR unblocks unit test with single Float8Linear module. Without the PR, we use `torch.uint8` only when dtypes are mixed. When passing float8 dtype to `foreach_copy_`, it throws error of `not being impelmented for float8 dtypes`

awgu replied:
Should we just support float8 dtypes in `_foreach_copy_`?

drisspg replied:
So we need all 4 for AMD support, at least the 2 e4 variants. I think Andrew's suggestion is right

weifengpy replied:
> Should we just support float8 dtypes in `_foreach_copy_`?

@awgu do you mean improving existing cuda kernels `foreach_tensor_copy_list_kernel_cuda_` to cover float8? or improve at python level?

awgu replied:
the former
my current understanding is that this should only require changing the dispatch macro for what dtypes are supported

weifengpy replied:
thanks for the pointers! I can give it a try

weifengpy committed:
Update on "[FSDP2] all-gather in uint8 for float8-only modules"


this PR unblocks unit test with single Float8Linear module. Without the PR, we use torch.uint8 only when dtypes are mixed. When passing float8 dtype to foreach_copy_, it throws error of not being impelmented for float8 dtypes

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented on aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:
```diff
@@ -285,44 +285,64 @@ struct Copy<dst_t, c10::complex<float>> {
-                                                      at::ScalarType::Bool,        \
-                                                      src_t,                       \
-                                                      __VA_ARGS__))
+#define AT_DISPATCH_SOURCE_TYPES(TYPE, NAME, ...)                                                \
```
I added Float8_e4m3fn, Float8_e4m3fnuz, Float8_e5m2, Float8_e5m2fnuz at the bottom of AT_DISPATCH_SWITCH. linter changed padded the entire AT_DISPATCH_SWITCH

vkuzo commented on aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:
```diff
@@ -410,10 +430,14 @@ void foreach_tensor_copy_list_kernel_cuda_(
   std::vector<std::vector<at::Tensor>> tensor_lists{src.vec(), self.vec()};
 
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND7(
```
if we use `AT_DISPATCH_V2` instead, would that help?

vkuzo replied:
I think changing to `AT_DISPATCH_V2` is low risk, but up to you.

weifengpy committed:
Update on "[FSDP2] support torch._foreach_copy_(float8) for fully_shard(Float8Linear)"


this PR unblocks unit test with single Float8Linear module. It fixes following error
```
torch._foreach_copy_(foreach_copy_dsts, all_gather_inputs)
[rank0]:E0913 13:44:29.829000 2179476 torch/testing/_internal/common_distributed.py:671] RuntimeError: "foreach_tensor_copy" not implemented for 'Float8_e4m3fn'
```


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented on test/distributed/_composable/fully_shard/test_fully_shard_util.py:
```diff
@@ -112,5 +121,20 @@ def test_get_sharded_module_tree_with_module_name_to_fqns(self):
+            torch.float8_e5m2,
+            torch.float8_e5m2fnuz,
+        ]:
+            src = [torch.zeros(2, 2, device="cuda", dtype=dtype)] * 2
```
I thought float8 is a sm8.9 thing. but do you mean we could offload float8 tensors to cpu, and this should test cpu as well?

eqy replied:
Additionally, is this not expected to work on `sm90`? Also `e5m2fnuz` might be a ROCM-only dtype
https://github.com/pytorch/pytorch/issues/126734#issuecomment-2333076704
These types of issues are hard to detect in open-source CI as the newest runners are only `sm86`

vkuzo replied:
@weifengpy , the float8 dtypes are hardware agnostic in PyTorch, I'd expect things like copy to work on cpu.  

Specific ops which take these dtypes as operands do have hardware restrictions (such as `torch._scaled_mm` on SM89+ on NVIDIA hardware).

weifengpy replied:
> e5m2fnuz might be a ROCM-only dtype

thanks for pointing it out

awgu commented on test/distributed/_composable/fully_shard/test_fully_shard_util.py:
```diff
@@ -112,5 +121,20 @@ def test_get_sharded_module_tree_with_module_name_to_fqns(self):
+            src = [torch.zeros(2, 2, device="cuda", dtype=dtype)] * 2
+            dst = [torch.zeros(2, 2, device="cuda", dtype=dtype)] * 2
+            # needed by fully_shard(Float8Linear)
+            torch._foreach_copy_(src, dst)
```
nit: we might as well go one step further and actually test that the copy was correct

e.g. change `src`/`dst` to rand values and then loop over them with `torch.equal`

weifengpy committed:
Update on "[FSDP2] support torch._foreach_copy_(float8) for fully_shard(Float8Linear)"


this PR unblocks unit test with single Float8Linear module. It fixes following error
```
torch._foreach_copy_(foreach_copy_dsts, all_gather_inputs)
[rank0]:E0913 13:44:29.829000 2179476 torch/testing/_internal/common_distributed.py:671] RuntimeError: "foreach_tensor_copy" not implemented for 'Float8_e4m3fn'
```


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge

PaliC commented:
@pytorchbot revert -m "breaking internal tests" -c "ghfirst"

weifengpy commented:
internal errors because of pytest version. will find some workaround

eqy commented on test/distributed/_composable/fully_shard/test_fully_shard_util.py:
```diff
@@ -112,5 +121,32 @@ def test_get_sharded_module_tree_with_module_name_to_fqns(self):
+            torch.float8_e4m3fn,
+            torch.float8_e4m3fnuz,
+            torch.float8_e5m2,
+            torch.float8_e5m2fnuz,
```
Is this a ROCM-only dtype?

weifengpy replied:
two of them (*uz) are for ROCM

weifengpy replied:
another two are for CUDA

eqy commented on test/distributed/_composable/fully_shard/test_fully_shard_util.py:
```diff
@@ -112,5 +121,32 @@ def test_get_sharded_module_tree_with_module_name_to_fqns(self):
 
 
+class TestUtilsSingleDevice(TestCase):
+    @pytest.mark.skipif(not is_cuda_8_9, reason="requires SM89 compatible machine")
```
Same as before, is it sm89+ or only sm89?

weifengpy replied:
this is sm89+

weifengpy replied:
i can rename to is_sm_89_plus if it helps

eqy replied:
In that case it might fit in here: https://github.com/pytorch/pytorch/blob/40b09edd87fcbe4e63c4db6399ec758d5c34e1b1/torch/testing/_internal/common_cuda.py#L27

weifengpy replied:
got you. will move there

weifengpy committed:
Update on "[FSDP2] support torch._foreach_copy_(float8) for fully_shard(Float8Linear)"


this PR unblocks unit test with single Float8Linear module. It fixes following error
```
torch._foreach_copy_(foreach_copy_dsts, all_gather_inputs)
[rank0]:E0913 13:44:29.829000 2179476 torch/testing/_internal/common_distributed.py:671] RuntimeError: "foreach_tensor_copy" not implemented for 'Float8_e4m3fn'
```


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented:
@weifengpy has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63961071).

weifengpy committed:
Update on "[FSDP2] support torch._foreach_copy_(float8) for fully_shard(Float8Linear)"


this PR unblocks unit test with single Float8Linear module. It fixes following error
```
torch._foreach_copy_(foreach_copy_dsts, all_gather_inputs)
[rank0]:E0913 13:44:29.829000 2179476 torch/testing/_internal/common_distributed.py:671] RuntimeError: "foreach_tensor_copy" not implemented for 'Float8_e4m3fn'
```


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D63961071](https://our.internmc.facebook.com/intern/diff/D63961071)

[ghstack-poisoned]

weifengpy commented:
@weifengpy has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63961071).

weifengpy commented:
fixed import error in interal import. landing this again

weifengpy commented:
@pytorchmergebot merge
